cli/src/main.rs

```diff
@@ -71,69 +77,88 @@ struct Opt {
+            cmac.copy_from_slice(&res);
             fs::write(opt.output_file, content)?;
         }
         AesGcm256Key::SCHEMA => {
             let iv = opt.iv.expect("IV is required.");
             let key = AesGcm256Key::new(&key, &iv)?;
             let mut content = fs::read(opt.input_file)?;
-            key.decrypt(&mut content)?;
+            let res = key.decrypt(&mut content)?;
+            cmac.copy_from_slice(&res);
             fs::write(opt.output_file, content)?;
         }
         TeaclaveFile128Key::SCHEMA => {
             let key = TeaclaveFile128Key::new(&key)?;
             let mut content = vec![];
-            key.decrypt(opt.input_file, &mut content)?;
+            let res = key.decrypt(opt.input_file, &mut content)?;
+            cmac.copy_from_slice(&res);
             fs::write(opt.output_file, content)?;
         }
         _ => bail!("Invalid crypto algorithm"),
     }
 
-    Ok(())
+    Ok(cmac)
 }
 
-fn encrypt(opt: EncryptDecryptOpt) -> Result<()> {
+fn encrypt(opt: EncryptDecryptOpt) -> Result<CMac> {
     let key = opt.key;
+    let mut cmac: CMac = [0u8; FILE_AUTH_TAG_LENGTH];
     match opt.algorithm.as_str() {
         AesGcm128Key::SCHEMA => {
             let iv = opt.iv.expect("IV is required.");
             let key = AesGcm128Key::new(&key, &iv)?;
             let mut content = fs::read(opt.input_file)?;
-            key.encrypt(&mut content)?;
+            let res = key.encrypt(&mut content)?;
+            cmac.copy_from_slice(&res);
             fs::write(opt.output_file, content)?;
         }
         AesGcm256Key::SCHEMA => {
             let iv = opt.iv.expect("IV is required.");
             let key = AesGcm256Key::new(&key, &iv)?;
             let mut content = fs::read(opt.input_file)?;
-            key.encrypt(&mut content)?;
+            let res = key.encrypt(&mut content)?;
+            cmac.copy_from_slice(&res);
             fs::write(opt.output_file, content)?;
         }
         TeaclaveFile128Key::SCHEMA => {
             let key = TeaclaveFile128Key::new(&key)?;
             let content = fs::read(opt.input_file)?;
-            key.encrypt(opt.output_file, &content)?;
+            let res = key.encrypt(opt.output_file, &content)?;
+            cmac.copy_from_slice(&res);
         }
         _ => bail!("Invalid crypto algorithm"),
     }
 
-    Ok(())
+    Ok(cmac)
 }
 
 fn main() -> Result<()> {
     let args = Opt::from_args();
-    match args.command {
-        Command::Decrypt(opt) => decrypt(opt)?,
-        Command::Encrypt(opt) => encrypt(opt)?,
-    }
+    let flag: bool;
+    let cmac = match args.command {
+        Command::Decrypt(opt) => {
+            flag = opt.cmac_flag;
+            decrypt(opt)?
+        }
+        Command::Encrypt(opt) => {
+            flag = opt.cmac_flag;
+            encrypt(opt)?
+        }
+    };
 
+    if flag {
+        let cmac_string = hex::encode(cmac);
+        println!("{}", cmac_string);
+    }
```

Comment:
You can move the this into `Command::Decrypt` and `Command::Encrypt`.

Since the CLI is not only designed for the decrypt/encrypt two sub-commands, there will be more other sub-commands which may have totally different arguments. Let's keep all logics only inside a sub-command branch.

I think it's ok to have two redundant cmac output code in these two sub command. What do you think?

Reply:
Thanks for the suggestion. Please see the latest commit.